### PR TITLE
Add default conformance to syntax collection

### PR DESCRIPTION
**Pull request by kimdv**
https://github.com/apple/swift/pull/40120

Added default conformance to syntax collection as discussed in #331

### CODE REVIEW

**Review by ahoppen — COMMENTED** [review]
LGTM. Will probably need some adjustment based on which changes will be made to https://github.com/apple/swift/pull/40120.

File: Sources/SwiftSyntaxBuilder/Buildables.swift.gyb [hunk 1]
```diff
@@ -42,8 +43,10 @@ public protocol ${kind}ListBuildable: SyntaxListBuildable {
 
 %     buildable_type = kind + 'Buildable'
 %     expressible_by_type = 'ExpressibleBy' + buildable_type
-%     expressible_by_protocols = SYNTAX_BUILDABLE_EXPRESSIBLE_BY_CONFORMANCES.get(expressible_by_type)
-%     if expressible_by_protocols:
+%     expressible_by_protocols = SYNTAX_COLLECTION_EXPRESSIBLE_BY_CONFORMANCES.get(buildable_type) or []
+%     expressible_by_protocols += SYNTAX_BUILDABLE_EXPRESSIBLE_BY_CONFORMANCES.get(buildable_type) or []
+%     expressible_by_protocols = map(lambda x : 'ExpressibleBy' +x, expressible_by_protocols)
```
Comment (ahoppen) [inline Sources/SwiftSyntaxBuilder/Buildables.swift.gyb:48]:
```suggestion
%     expressible_by_protocols = map(lambda x : 'ExpressibleBy' + x, expressible_by_protocols)
```

File: Sources/SwiftSyntaxBuilder/Buildables.swift.gyb [hunk 2]
```diff
@@ -42,8 +43,10 @@ public protocol ${kind}ListBuildable: SyntaxListBuildable {
...
+%     if expressible_by_protocols and expressible_by_protocols is not []:
```
Comment (ahoppen) [inline Sources/SwiftSyntaxBuilder/Buildables.swift.gyb:49]: I think 
```suggestion
%     if expressible_by_protocols:
```
should be enough.

File: Sources/SwiftSyntaxBuilder/Buildables.swift.gyb [hunk 3]
```diff
@@ -220,8 +223,10 @@ public struct ${node.syntax_kind}: SyntaxBuildable {
 %   end
 %   if node.is_buildable() or node.is_syntax_collection():
 %     expressible_by_type = 'ExpressibleBy' + node.syntax_kind
-%     expressible_by_protocols = SYNTAX_BUILDABLE_EXPRESSIBLE_BY_CONFORMANCES.get(expressible_by_type)
-%     if expressible_by_protocols:
+%     expressible_by_protocols = SYNTAX_COLLECTION_EXPRESSIBLE_BY_CONFORMANCES.get(node.syntax_kind) or []
+%     expressible_by_protocols += SYNTAX_BUILDABLE_EXPRESSIBLE_BY_CONFORMANCES.get(node.syntax_kind) or []
+%     expressible_by_protocols = map(lambda x : 'ExpressibleBy' +x, expressible_by_protocols)
```
Comment (ahoppen) [inline Sources/SwiftSyntaxBuilder/Buildables.swift.gyb:228]:
```suggestion
%     expressible_by_protocols = map(lambda x : 'ExpressibleBy' + x, expressible_by_protocols)
```

File: Sources/SwiftSyntaxBuilder/Buildables.swift.gyb [hunk 4]
```diff
@@ -249,37 +254,40 @@ extension TokenSyntax: ExpressibleByTokenSyntax {
...
 }
 
 %   end
-%   if 'ExpressibleByConditionElement' in conformances:
-extension ${protocol} {
+% end
+
+% for protocol, conformances in SYNTAX_BUILDABLE_EXPRESSIBLE_BY_CONFORMANCES.items():
+%   if 'ConditionElement' in conformances:
+extension ExpressibleBy${protocol} {
```
Comment (ahoppen) [inline Sources/SwiftSyntaxBuilder/Buildables.swift.gyb:270]: Not part of this PR, but couldn’t we also generate these `extension` implementations now that syntax collection conformances live somewhere else?

Comment (kimdv) [reply Sources/SwiftSyntaxBuilder/Buildables.swift.gyb:270]: I would be nice! 
But the thing is, that the init method different for each type. 

We can enrich the `protocols.py`file with the needed information.

Comment (ahoppen) [reply Sources/SwiftSyntaxBuilder/Buildables.swift.gyb:270]: Why is that? Am I missing something obvious? Shouldn’t the following work?

```swift
extension ExpressibleBy${protocol} {
   public func create${expressible_as_protocol}() -> ${expressible_as_protocol} {
     ${expressible_as_protocol}(${retrieve only non-optional child name here}: self)
   }
 }

```

Comment (kimdv) [reply Sources/SwiftSyntaxBuilder/Buildables.swift.gyb:270]:
```swift
extension ExpressibleByDeclBuildable {
  public func createCodeBlockItem() -> CodeBlockItem {
    CodeBlockItem(item: self)
  }
}

extension ExpressibleByDeclBuildable {
  public func createMemberDeclListItem() -> MemberDeclListItem {
    MemberDeclListItem(decl: self)
  }
}
```

Yes. Sorry bad explanation 😅
The `${retrieve only non-optional child name here}` do we need to find somehow. 
Not sure how to do, but I will look into it, and then we can take it from there. 

It is possible, somehow 😬

**Review by ahoppen — APPROVED** [review]
Looking good. Could you rebase because I just reverted my change of `ExpressibleAs` -> `ExpressibleBy` because I had a brain fart yesterday. The protocols actually describe that these types are expressible **as** other types and not **by** other types. Also make sure to rename all of your `by`s by `as`s.

**Comment by kimdv** [comment]
I saw that! 

I have rebased 🚀

**Comment by kimdv** [comment]
apple/swift#40120

@swift-ci Please test

**Comment by ahoppen** [comment]
apple/swift#40120

@swift-ci Please test
